web/source/keyboards/kmwkeyboards.ts

```diff
@@ -1129,34 +1134,63 @@ namespace com.keyman.keyboards {
         }
       } else if(options['context'] == 'language') { // Download the full list of supported keyboard languages
         this.languageList = x['languages'];
-        if(this.languagesPending) {
-          this.addLanguageKeyboards(this.languagesPending);
-        }
-        this.languagesPending = [];
       }
 
-      return null; // TODO (#5044): return the list of newly-registered stubs here.
+      return this.keyboardStubs.slice(currentKeyboardStubsCount);
     }
 
     /**
      *  Internal handler for processing keyboard registration, used only by `register`
      *
-     *  @param  {Object}   languages    Array of language names
+     *  @param  {string[]}   languages    Array of language names
+     *  @returns {Promise<KeyboardStub[]>} Promise of added keyboard stubs
      **/
-    addLanguageKeyboards(languages) {
-      var i, j, lgName, cmd, first, addAll;
+    addLanguageKeyboards(languages: string[]): Promise<KeyboardStub[]> {
+      var i, j, lgName, cmd, addAll, retPromise;
 
       // Defer registering keyboards by language until the language list has been loaded
-      if(this.languageList == null) {
-        first = (this.languagesPending.length == 0);
-
-        for(i=0; i<languages.length; i++) {
-          this.languagesPending.push(languages[i]);
+      if (this.languageList == null) {
+        if (this.firstCall) {
+          this.firstCall = false;
+          let promise = this.keymanCloudRequest('',true);
+          // If promise is not error, then... (needs an error guard)
+          promise.catch(function(error) {
+            console.log('first error', error);
+            return Promise.reject(error);
+          });
+          this.languageListPromise = promise;
         }
 
-        if(first) {
-          this.keymanCloudRequest('',true);
-        }
+        let _this = this;
+        retPromise = new Promise(function(resolve, reject) {
+          if (_this.languageListPromise) {
+            // 1: wait for the language list to be loaded properly
+            _this.languageListPromise.then(function() {
+              // 2: perform the actual query, now that we can find the language code
+              let innerPromise = _this.addLanguageKeyboards(languages);
+
+              // 3: intercept the Promise resolve/reject and pass them to the
+              // returned Promise.
+              innerPromise.then(function(stubs) {
+                resolve(stubs);
+              }).catch(function(error) {
+                reject(error);
+              });
```

Comment:
Sorry I only realized this later, but...

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve

We can actually replace all of this with:

```suggestion
              resolve(_this.addLanguageKeyboards(languages));
```

In short, `Promise`s can 'resolve' on inner `Promise`s.